pkg/metrics/node.go

```diff
@@ -0,0 +1,194 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/rancher/wrangler/pkg/ticker"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	v32 "github.com/rancher/rancher/pkg/apis/management.cattle.io/v3"
+	mgmtcontrollers "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	v3 "github.com/rancher/rancher/pkg/generated/norman/management.cattle.io/v3"
+)
+
+const (
+	nodeProviderLabel  = "provider"
+	nodeClusterIDLabel = "cluster_id"
+)
+
+var (
+	nodeLabels = []string{nodeProviderLabel, nodeClusterIDLabel}
+	numNodes   = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: "cluster_manager",
+			Name:      "nodes",
+			Help:      "Number of nodes in ready clusters",
+		}, nodeLabels,
+	)
+	numCores = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Subsystem: "cluster_manager",
+			Name:      "node_cores",
+			Help:      "Number of node cores in ready clusters",
+		}, nodeLabels,
+	)
+)
+
+const (
+	nodePollInterval = time.Duration(15 * time.Second)
+)
+
+type nodeMetrics struct {
+	nodeCache    mgmtcontrollers.NodeCache
+	clusterCache mgmtcontrollers.ClusterCache
+}
+
+func (m *nodeMetrics) collect(ctx context.Context) {
+	for range ticker.Context(ctx, nodePollInterval) {
+		logrus.Debugf("[node-metrics] collecting nodes to report metrics")
+
+		nodes, err := m.listNodes()
+		if err != nil {
+			logrus.Errorf("[node-metrics] couldn't list v3.Nodes: %v", err)
+			continue
+		}
+
+		var infos []*nodeInfo
+		for _, node := range nodes {
+			info, err := m.getNodeInfo(node)
+			if err != nil {
+				logrus.Debugf("[node-metrics] could not determine node info: %v", err)
+				continue
+			}
+			infos = append(infos, info)
+		}
+
+		setMetrics(infos)
+	}
+
+	logrus.Debugf("[node-metrics] context cancelled, exiting")
+}
+
+func (m *nodeMetrics) listNodes() ([]*v3.Node, error) {
+	var nodes []*v3.Node
+	backoff := wait.Backoff{
+		Duration: 500 * time.Millisecond,
+		Factor:   1.5,
+		Steps:    10,
+		Cap:      time.Duration(30 * time.Second),
+	}
+
+	err := wait.ExponentialBackoff(backoff, func() (done bool, err error) {
+		nodes, err = m.nodeCache.List("", labels.NewSelector())
+		if err != nil {
+			return false, nil // retry on error, may need to limit to certain types of errors
+		}
+
+		// retry if there are no nodes
+		if len(nodes) == 0 {
+			return false, nil
+		}
+
+		return true, nil
+	})
+
+	return nodes, err
+}
+
+type nodeLabelValues struct {
+	provider  string
```

Comment:
Yea, so the metrics themselves are labeled by `provider` and `cluster_id`, this is because downstream of the metrics we need to know how many nodes there are per provider, example of a provider label values.
```
gke
aks
eks
rke
rke2
rke.windows
```
The `cluster_id` is there to be able to distinguish management nodes from downstream nodes

Reply:
Ok - so the provider is something that we need to meet the requirements as outlined in the issue. Thanks for explaining.